### PR TITLE
MOSART-inundation restart fix and other minor bug fixes

### PR DESCRIPTION
**Pull request by hydrotian**
This PR consists of multiple bug fixes and improvements:
- Added variables relative to MOSART-inundation into MOSART restart file.
- Fixed a restart bug for MOSART-WM so that the model can generate BFB results when simulation is less than a month. 
- Fixed a bug for dam construction in MOSART-WM.
- Added irrigation term to ELM water budget report.
- Revert some changes in ELM to allow irrigation when crop model is off.

[non-BFB] only for SMS.MOS_USRDAT.RMOSGPCC._.mosart-unstructure

### CODE REVIEW

**Comment by hydrotian** [comment]
@bishtgautam I have done the 25 land development tests with only one failed (SMS.MOS_USRDAT.RMOSGPCC.compy_intel.mosart-unstructure) because this test has inundation turned on. My PR changed the answer of inundation so I'm not sure if it can still be called BFB.

**Comment by rljacob** [comment]
I assume this will have to wait until the v2 water cycle is tagged.  @golaz ?

**Comment by bishtgautam** [comment]
@rljacob The single test that is failing has inundation turned on in MOSART. In watercycle V2, inundation is off. Thus, this PR is BFB w.r.t. to watercycle V2.

File: components/elm/src/main/controlMod.F90 [hunk 1]
```diff
@@ -445,10 +445,10 @@ subroutine control_init( )
...
        end if
        
-       if (.not. use_crop .and. irrigate) then
-          call endrun(msg=' ERROR: irrigate = .true. requires CROP model active.'//&
-            errMsg(__FILE__, __LINE__))
-       end if
+       !if (.not. use_crop .and. irrigate) then
+       !   call endrun(msg=' ERROR: irrigate = .true. requires CROP model active.'//&
+       !     errMsg(__FILE__, __LINE__))
+       !end if
```
Comment (bishtgautam) [inline components/elm/src/main/controlMod.F90:451]: Please delete these lines

**Review by bishtgautam — COMMENTED** [review]
@hydrotian A minor change is needed.

File: components/mosart/src/riverroute/RtmRestFile.F90 [hunk 2]
```diff
@@ -445,6 +446,48 @@ subroutine RtmRestart(ncid, flag)
...
+          dfld  => rtmCTL%inundhf(:)
+       elseif (nv == 11 .and. trim(rtm_tracers(nt)) == 'LIQ') then
+          vname = 'RTM_INUNDFF_'//trim(rtm_tracers(nt))
+          lname = 'inundation fp area fraction'
+          uname = 'no unit'
+          dfld  => rtmCTL%inundff(:)        
+       elseif (nv == 12 .and. trim(rtm_tracers(nt)) == 'LIQ') then
+          vname = 'RTM_MR_'//trim(rtm_tracers(nt))
+          lname = 'mr'
+          uname = 'no unit'
```
Comment (liho745) [inline components/mosart/src/riverroute/RtmRestFile.F90:471]: It has a unit [m2]

File: components/mosart/src/riverroute/RtmRestFile.F90 [hunk 3]
```diff
@@ -445,6 +446,48 @@ subroutine RtmRestart(ncid, flag)
...
+          dfld  => rtmCTL%mr(:,nt)   
+       elseif (nv == 13 .and. trim(rtm_tracers(nt)) == 'LIQ') then
+          vname = 'RTM_YR_'//trim(rtm_tracers(nt))
+          lname = 'yr'
+          uname = 'no unit'
```
Comment (liho745) [inline components/mosart/src/riverroute/RtmRestFile.F90:476]: It has a unit [m]

File: components/mosart/src/riverroute/RtmRestFile.F90 [hunk 4]
```diff
@@ -445,6 +446,48 @@ subroutine RtmRestart(ncid, flag)
...
+          dfld  => rtmCTL%yr(:,nt)   
+       elseif (nv == 14 .and. trim(rtm_tracers(nt)) == 'LIQ') then
+          vname = 'RTM_PR_'//trim(rtm_tracers(nt))
+          lname = 'pr'
+          uname = 'no unit'
```
Comment (liho745) [inline components/mosart/src/riverroute/RtmRestFile.F90:481]: It has a unit [m]

File: components/mosart/src/riverroute/RtmRestFile.F90 [hunk 5]
```diff
@@ -445,6 +446,48 @@ subroutine RtmRestart(ncid, flag)
...
+          dfld  => rtmCTL%pr(:,nt)   
+       elseif (nv == 15 .and. trim(rtm_tracers(nt)) == 'LIQ') then
+          vname = 'RTM_RR_'//trim(rtm_tracers(nt))
+          lname = 'rr'
+          uname = 'no unit'
```
Comment (liho745) [inline components/mosart/src/riverroute/RtmRestFile.F90:486]: It has a unit [m]

File: components/mosart/src/riverroute/RtmRestFile.F90 [hunk 6]
```diff
@@ -536,12 +607,17 @@ subroutine RtmRestart(ncid, flag)
              if (abs(rtmCTL%wt(n,nt))      > 1.e30) rtmCTL%wt(n,nt) = 0.
              if (abs(rtmCTL%wr(n,nt))      > 1.e30) rtmCTL%wr(n,nt) = 0.
              if (abs(rtmCTL%erout(n,nt))   > 1.e30) rtmCTL%erout(n,nt) = 0.
+             if (abs(rtmCTL%inundffunit(n))> 1.e30) rtmCTL%inundffunit(n) = 0.
+             if (abs(rtmCTL%inundwf(n))    > 1.e30) rtmCTL%inundwf(n) = 0.
+             if (abs(rtmCTL%inundhf(n))    > 1.e30) rtmCTL%inundhf(n) = 0.
+             if (abs(rtmCTL%inundff(n))    > 1.e30) rtmCTL%inundff(n) = 0.
```
Comment (liho745) [inline components/mosart/src/riverroute/RtmRestFile.F90:613]: Do you want to add similar lines for rtmCLT%mr, rtmCLT%yr, etc.?

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 7]
```diff
@@ -1595,6 +1595,10 @@ subroutine Rtmini(rtm_active,flood_active)
        TRunoff%wh   = rtmCTL%wh
        TRunoff%wt   = rtmCTL%wt
        TRunoff%wr   = rtmCTL%wr
+       TRunoff%mr   = rtmCTL%mr
+       TRunoff%yr   = rtmCTL%yr
+       TRunoff%pr   = rtmCTL%pr
+       TRunoff%rr   = rtmCTL%rr
```
Comment (liho745) [inline components/mosart/src/riverroute/RtmMod.F90:1601]: It is not quite clear to me why we need these channel cross-section geometry values. In principle, they can all be derived from the channel storage by invoking the updatestate_mainchannel subroutine.

**Comment by bishtgautam** [comment]
@hydrotian This PR is causing DIFF for a river-only compset. See https://my.cdash.org/test/31430949.

**Comment by rljacob** [comment]
@hydrotian is the diff from baselines in SMS.MOS_USRDAT.RMOSGPCC.*_*.mosart-unstructure  expected?

**Comment by rljacob** [comment]
Sorry, I was looking at the [BFB] note in the PR.  If any test in integration diffs, its a non-BFB PR.   @bishtgautam you can bless that diff and proceed.

**Comment by bishtgautam** [comment]
Tickets have been filed to bless the DIFFs.

**Pull request merged**